packages/editor/src/editor.ts

```diff
@@ -0,0 +1,236 @@
+  // we should detect whether the MongoDB extension is installed and open a .mongodb file.
+  // In all other cases we should open a .js file.
+  async _getExtension(cmd: string): Promise<string> {
+    if (!this._isVscodeApp(cmd)) {
+      return 'js';
+    }
+
+    try {
+      const extensions = await fse.readdir(path.join(this._vscodeDir, 'extensions'));
+      const hasMongodbExtension = !!extensions
+        .find((name) => name.includes('mongodb.mongodb-vscode'));
+
+      this.messageBus.emit('mongosh-editor:read-vscode-extensions-done', {
+        vscodeDir: this._vscodeDir,
+        hasMongodbExtension
+      });
+
+      return hasMongodbExtension ? 'mongodb' : 'js';
+    } catch (error) {
+      this.messageBus.emit('mongosh-editor:read-vscode-extensions-failed', {
+        action: 'mongosh-editor:read-vscode-extensions-failed',
+        error: error.message
+      });
+
+      return 'js';
+    }
+  }
+
+  async _getEditor(): Promise<string|null> {
+    // Check for an external editor in the mongosh configuration.
+    let editor: string | null = await this._instanceState.shellApi.config.get('editor');
+
+    // Check for an external editor in the environment variable.
+    if (!editor && process.env.EDITOR) {
+      editor = process.env.EDITOR;
+    }
+
+    return editor;
+  }
+
+  async _createTempFile(): Promise<void> {
+    const ext = await this._getExtension(this._editorSession.cmd);
+    const content = await this._getEditorContent();
+
+    this._editorSession.tmpDoc = path.join(this._tmpDir, `edit-${uuidv4()}.${ext}`);
+
+    // Create a temp file to store content that is being edited.
+    await fse.ensureFile(this._editorSession.tmpDoc);
+    await fse.writeFile(this._editorSession.tmpDoc, beautify(content));
+  }
+
+  async _readTempFile(tmpDoc: string): Promise<string> {
+    const content = await fse.readFile(tmpDoc, 'utf8');
+    await fse.unlink(tmpDoc);
+    return this._makeMultilineJSIntoSingleLine(content);
+  }
+
+  _isVscodeApp(cmd: string): boolean {
+    const regex = /^((.*[\/\\])*)?[cC]ode(.exe)?$/;
+
+    try {
+      return !!regex.exec(cmd);
+    } catch (error) {
+      this.messageBus.emit('mongosh-editor:is-vscode-app-regex-failed', { regex, cmd });
+      return false;
+    }
+  }
+
+  _isStatement(code: string): boolean {
+    const regex = /\b[^()]+\((.*)\)$/;
+
+    try {
+      return !!regex.exec(code);
+    } catch (error) {
+      this.messageBus.emit('mongosh-editor:is-statement-regex-failed', { regex, code });
+      return false;
+    }
+  }
+
+  async _getEditorContent(): Promise<string> {
+    if (!this._editorSession.code) {
+      return this._editorSession.content;
+    }
+
+    const evalResult = await this._loadExternalCode(this._editorSession.code, '@(shell eval)');
+    return this._isStatement(this._editorSession.code) ? this._editorSession.code : evalResult.toString();
+  }
+
+  async runEditCommand([ code, ...args ]: string[]): Promise<void> {
```

Comment:
We also need to be able to run things like `edit function foo() { return 'a  b'; }`.

The arguments here would currently be split at whitespace – that’s not useful for us, because it loses information (e.g. about neighboring whitespaces). We’ll need to be able to pass a single string from the shell-evaluator package to this function.